internal/workerpool/pool.go

```diff
@@ -20,32 +20,46 @@ package workerpool
 
 import (
 	"context"
+	"sync"
 )
 
-type Task func(context.Context)
+type Task func(context.Context) error
 
 // Run runs a worker pool with no more than limit goroutines. It repeatedly
 // calls nextTask to get tasks to perform. These tasks are in the form of funcs
 // that do some work via side effects. If nextTask returns nil, Run will no
 // longer ask for more tasks, and will return after waiting for the running
 // goroutines to finish. The provided context can be used to cancel everything
 // and exit the loop.
-func Run(ctx context.Context, limit int, nextTask func(context.Context) Task) {
+func Run(ctx context.Context, limit int, nextTask func(context.Context) (Task, error)) error {
+	ctx, cancel := context.WithCancel(ctx)
 	type token struct{}
 	sem := make(chan token, limit)
+	var mu sync.Mutex
+	var retErr error
+	setErr := func(err error) {
+		mu.Lock()
+		retErr = err
```

Comment:
What happened to `if retErr == nil`? I think you want that to capture the first error.

Reply:
Done.